File: src/main/java/rs/raf/javaproject/model/SuccessorTable.java
package rs.raf.javaproject.model;

import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import rs.raf.javaproject.repository.Database;

import java.util.ArrayList;
import java.util.Collection;

@Data
@Component
public class SuccessorTable {

    @Autowired
    private Database database;

    private final ArrayList<Node> table;

    public void reconstructTable(){
        table.clear();
        ArrayList<Node> list = new ArrayList<>(database.getAllNodes().values());

        int myPos = list.indexOf(database.getInfo());

        int size = list.size();
        for (int step = 1; step < size; step *=2){

            int succPos = (myPos + step) % size;
            table.add(list.get(succPos));
        }
        System.out.println(database.getInfo().getId() + ":" + table);
    }

    public Collection<Node> broadcastingNodes(){
        ArrayList<Node> broadcastingNodes = new ArrayList<>();

        try {
            broadcastingNodes.add(table.get(0));
            broadcastingNodes.add(table.get(1));
            broadcastingNodes.add(table.get(2));
        }catch (IndexOutOfBoundsException e){
        }

        return broadcastingNodes;
    }

    public Node getDelegator(Node node){

        if(table.size() == 0) return null; // Cisto provere radi

        Node ret = table.get(0);

        for(Node tableNode : table){
            if(node.compareTo(tableNode) == -1){
                break;
            }
            ret = tableNode;
        }

        return ret;
    }
}

File: model/SuccessorTable.py
from repository.Database import Database


class SuccessorTable:
    pass

    __instance = None

    database = Database.getInstance()
    table = []

    @staticmethod
    def getInstance():
        if SuccessorTable.__instance == None:
            SuccessorTable()
        return SuccessorTable.__instance

    def __init__(self):
        if SuccessorTable.__instance != None:
            raise  Exception("This class is singleton!")
        else:
            SuccessorTable.__instance = self


    def reconstructTable(self):
        self.table = []

        tmpList = list(self.database.allNodes.values())

        myPos = tmpList.index(self.database.getInfo())

        size = len(tmpList)

        for step in range(1, size):
            succPos = (myPos + step) % size
            self.table.append(tmpList[succPos])
            step *= 2

        print("Successor table: ", [e.getID() for e in self.table])

    def getBroadcastingNodes(self):
        broadcastingNodes = []

        try:
            broadcastingNodes.append(self.table[0])
            broadcastingNodes.append(self.table[1])
            broadcastingNodes.append(self.table[2])
        except IndexError:
            pass

        return broadcastingNodes